### ISSUE TITLE
How to properly move submodules to GPU?

### ISSUE THREAD

**Issue by MatthewMatero** [PyTorchLightning/pytorch-lightning#1310]
I've coded up a TransformerEncoder that relies on submodules. Specifically, I have a main lightning module (MainTransfomer.py) which has 2 sub (regular torch)modules. 1 is BertModel and 1 is a custom TransformerEncoderLayer. 

The TransformerEncoderLayer has 2 submodules of its own named MultiHeadAttn and PosFeedForward. When I try to run my code via the lightning trainer, it seems that all my tensors/params are being moved to GPU except for when it tries to do operations within the MultiHeadAttn and PosFeedForward modules. Specifically it seems the weight matrix of a nn.Linear() within MultiHeadAttn is still sitting on cpu. 

So my question boils down to what is the correct way to ensure all submodules are moved to the (correct) GPU within the lightning framework? In normal PyTorch I would just explicitly call .to(device) but I've read that it is not recommended to do this within lightning. 

if I explicitly set .to(device) on my TransformerEncoderLayer in my MainTransformerEncoder __init__ I don't run into this issue. 

I can supply code if need be, but the general setup is as described above. Main lightning module inits a torch module, which itself has 2 sub torch modules (attn & ffnn). The attn and ffnn modules don't seem to be moved to GPU by the lightning trainer.

Env:
 - Ubuntu16.04
 - conda/py3.8
 - Lightning v0.7.1

**Comment by awaelchli** [comment]
Strange. Lightning should move all your modules to the GPU given they are LightningModules or are part of one. To identify and fix a potential bug we would need a minimal example. Could you post it as a google colab notebook? That would be super helpful.

**Comment by MatthewMatero** [comment]
Hey, I tried moving my code into collab this morning. I tried to find a balance between leaving the code as is and putting in the neccessary bits to just show the error. In the colab notebook I've confirmed the code runs beyond the point of the original error when using CPU. However, when switching to GPU within the colab I'm getting a vague error of "CUDA error: An illegal memory access was encountered".

Here's the notebook containing the code: [link](https://colab.research.google.com/drive/1Na7_2dyQkeFTv9WaiPya3yYvh7nHxMjX#scrollTo=zApcmeVkK58a)
